os/src/SubProcess.scala

```diff
@@ -339,28 +353,40 @@ class ProcessPipeline(
     }
   }
 
+  // FIXME: documentation
   /**
    * Wait up to `millis` for the [[ProcessPipeline]] to terminate all the processes
    * in pipeline. By default waits indefinitely; if a time limit is given, explicitly
    * destroys each process if it has not completed by the time the timeout has occurred.
    */
-  override def join(timeout: Long = -1): Boolean = {
-    @tailrec
-    def joinRec(startedAt: Long, processesLeft: Seq[SubProcess], result: Boolean): Boolean =
-      processesLeft match {
-        case Nil => result
-        case head :: tail =>
-          val elapsed = System.currentTimeMillis() - startedAt
-          val timeoutLeft = Math.max(0, timeout - elapsed)
-          val exitedCleanly = head.join(timeoutLeft)
-          joinRec(startedAt, tail, result && exitedCleanly)
-      }
+  override def join(timeout: Long = -1, timeoutGracePeriod: Long = 1000): Boolean = {
+    // previously, this was implemented in a similar way to the waitFor above, but with
+    // join logic. This is much harder to make work with the grace period, because we
+    // want to evenly give all threads a chance to terminate gracefully. As such, this
+    // implementation interleaves the single-process join implementation more fairly
 
+    // in this case, the grace period does not apply, so fine
     if (timeout == -1) {
       processes.forall(_.join())
     } else {
-      val timeNow = System.currentTimeMillis()
-      joinRec(timeNow, processes, true)
+      // timeout is active, so the grace period must be accounted for
+      val exitedCleanly = waitFor(timeout)
+      if (!exitedCleanly) {
+        if (timeoutGracePeriod == -1) destroy()
+        else if (timeoutGracePeriod == 0) destroyForcibly()
+        else {
+          destroy()
+          if (!waitFor(timeoutGracePeriod)) {
+            destroyForcibly()
+          }
+        }
+        waitFor(-1)
+        // note that this is the only part that isn't shared with the other implementation of join... they could be unified if this is made an
+        // abstract method: then this implementation can move to the superclass (except for the default -1 case above, but that could be done via an override)
+        processes.flatMap(_.outputPumperThread).foreach(_.join())
```

Comment:
These are the only bit of this implementation that isn't shared with the single process version. If `joinPumperThreads` was created as a new `protected` abstract method on the super-trait, then both implementations of `join` can be cleanly factored out as a default implementation in the trait itself, reducing code repetition.

Thoughts?